src/util/BuiltInFunctions.lua

```diff
@@ -0,0 +1,104 @@
+--Copyright 2021 The casbin Authors. All Rights Reserved.
+--
+--Licensed under the Apache License, Version 2.0 (the "License");
+--you may not use this file except in compliance with the License.
+--You may obtain a copy of the License at
+--
+--    http://www.apache.org/licenses/LICENSE-2.0
+--
+--Unless required by applicable law or agreed to in writing, software
+--distributed under the License is distributed on an "AS IS" BASIS,
+--WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--See the License for the specific language governing permissions and
+--limitations under the License.
+
+local rex = require ("rex_pcre")
+local posix = require("posix.fnmatch")
+
+BuiltInFunctions = {}
+
+-- KeyMatch determines whether key1 matches the pattern of key2 (similar to RESTful path), key2 can contain a *.
+-- For example, "/foo/bar" matches "/foo/*"
+function BuiltInFunctions.keyMatch(key1, key2)
+    local i, _ = string.find(key2, "*")
+
+    if not i then
+        return (key1 == key2)
+    end
+
+    if #key1>=i then
+        return (string.sub(key1, 1, i-1) == string.sub(key2, 1, i-1))
+    end
+    return (key1 == string.sub(key2, 1, i-1))
+end
+
+-- KeyGet returns the matched part
+-- For example, "/foo/bar/foo" matches "/foo/*"
+-- "bar/foo" will been returned
+function BuiltInFunctions.keyGet(key1, key2)
+    local i, _ = string.find(key2, "*")
+
+    if not i then
+        return ""
+    end
+    if #key1>=i then
+        if string.sub(key1, 1, i-1) == string.sub(key2, 1, i-1)  then
+            return string.sub(key1, i)
+        end
+    end
+    return ""
+end
+
+-- KeyMatch2 determines whether key1 matches the pattern of key2 (similar to RESTful path), key2 can contain a *.
+-- For example, "/foo/bar" matches "/foo/*", "/resource1" matches "/:resource"
+function BuiltInFunctions.keyMatch2(key1, key2)
+    key2 = string.gsub(key2, "/%*", "/.*")
+    local key = rex.gsub(key2, ":[^/]+", "[^/]+")
+	return BuiltInFunctions.regexMatch(key1, "^"..key.."$")
+end
+
+-- KeyMatch3 determines whether key1 matches the pattern of key2 (similar to RESTful path), key2 can contain a *.
+-- For example, "/foo/bar" matches "/foo/*", "/resource1" matches "/{resource}"
+function BuiltInFunctions.keyMatch3(key1, key2)
+    key2 = string.gsub(key2, "/%*", "/.*")
+    local key = rex.gsub(key2, "{[^/]+}", "[^/]+")
+	return BuiltInFunctions.regexMatch(key1, "^"..key.."$")
+end
+
+-- RegexMatch determines whether key1 matches the pattern of key2 in regular expression.
+function BuiltInFunctions.regexMatch(key1, key2)
+    local res = rex.match(key1, key2)
+    if res then
+        return true
+    else
+        return false
+    end
+end
+
+-- GlobMatch determines whether key1 matches the pattern of key2 using glob pattern
+function BuiltInFunctions.globMatch(key1, key2)
+    if posix.fnmatch(key2, key1) == 0 then
```

Comment:
```
 posix.fnmatch(key2, key1, posix.FNM_PATHNAME | posix.FNM_PERIOD) 
```
Only need to specify flags ([Module posix.fnmatch](http://luaposix.github.io/luaposix/modules/posix.fnmatch.html)), the result will be the same as Go .

Reply:
@techoner I have added this flag and now it works similar to Casbin (Go).